lib/Dialect/SPIRV/Transforms/DecorateWithLayoutInfoSPIRVPass.cpp

```diff
@@ -0,0 +1,282 @@
+}
+
+Type SPIRVLayoutInfoUtils::decorateType(spirv::CompositeType compositeType,
+                                  spirv::StructType::LayoutInfo &offset) {
+  if (!compositeType.getNumElements()) {
+    return compositeType;
+  }
+  uint64_t elementSizeInBytes = 0;
+  auto memberType = SPIRVLayoutInfoUtils::decorateType(
+      compositeType.getElementType(0), elementSizeInBytes);
+  offset += elementSizeInBytes * compositeType.getNumElements();
+  return SPIRVLayoutInfoUtils::getArrayOrVectorType(compositeType, memberType,
+                                                    elementSizeInBytes);
+}
+
+Type SPIRVLayoutInfoUtils::getArrayOrVectorType(
+    spirv::CompositeType compositeType, Type elementType,
+    uint64_t arrayStride) {
+  switch (compositeType.getKind()) {
+  case spirv::TypeKind::Array:
+    return spirv::ArrayType::get(elementType, compositeType.getNumElements(),
+                                 arrayStride);
+  case StandardTypes::Vector:
+    return VectorType::get(compositeType.getNumElements(), elementType);
+  default:
+    llvm_unreachable("unhandled SPIR-V composite type");
+  }
+}
+
+unsigned SPIRVLayoutInfoUtils::getPrimitiveTypeSizeInBytes(Type primitiveType) {
+  // According to the SPIR-V spec:
+  // "The ArrayStride, MatrixStride, and Offset decorations must be large
+  // enough to hold the size of the objects they affect (that is, specifying
+  // overlap is invalid)."
+  auto bitWidth = primitiveType.getIntOrFloatBitWidth();
+  if (bitWidth == 1 || bitWidth == 8 || bitWidth == 16)
+    return 4;
+  return bitWidth / 8;
+}
+
+bool SPIRVLayoutInfoUtils::isLegalType(Type type) {
+  auto ptrType = type.dyn_cast<spirv::PointerType>();
+  if (!ptrType) {
+    return true;
+  }
+  auto storageClass = ptrType.getStorageClass();
+  auto structType = ptrType.getPointeeType().dyn_cast<spirv::StructType>();
+  if (!structType) {
+    return true;
+  }
+  if (storageClass != spirv::StorageClass::Uniform &&
+           storageClass != spirv::StorageClass::StorageBuffer &&
+           storageClass != spirv::StorageClass::PushConstant) {
+    return true;
+  }
+  return structType.hasLayout();
+}
+
+bool SPIRVLayoutInfoUtils::isPrimitiveType(Type primitiveType) {
+  if (auto integerType = primitiveType.dyn_cast<IntegerType>()) {
+    return llvm::is_contained(llvm::ArrayRef<unsigned>({1, 8, 16, 32, 64}),
+                              integerType.getWidth());
+  }
+  return primitiveType.isF32() || primitiveType.isF64();
+}
+
+namespace {
+class SPIRVGlobalVariableOpLayoutInfoDecoration
+    : public OpRewritePattern<spirv::GlobalVariableOp> {
+public:
+  using OpRewritePattern<spirv::GlobalVariableOp>::OpRewritePattern;
+
+  PatternMatchResult matchAndRewrite(spirv::GlobalVariableOp op,
+                                     PatternRewriter &rewriter) const override {
+    // Initial offset is zero.
+    spirv::StructType::LayoutInfo baseOffset = 0;
```

Comment:
If the struct already has the layout information, nothing should be done (match should fail)

Reply:
You are right, but the check is under isLegalType, in terms of legalization I assume that type is illegal if it's a !spv.ptr<!spv.struct<...> in strorage class StorageBuffer/Uniform/PushConstant and does not have layout info, otherwise it could be illegal type with layout info, which is wrong. So, If type has layout info the pattern is not applicable here, I've provided a test with custom layout info. Do I miss something? Please fix me, if I'm wrong.

Reply:
I am not sure when the `isLegal` function is invoked during the conversion. You might be right. But do add a test where the layout of the struct is different from what is assigned by default in this pass, and that shouldn't change. If that test passes, then I guess this is fine.

Reply:
isLegalType function is invoked:
```
  target.addDynamicallyLegalOp<spirv::GlobalVariableOp>(
      [](spirv::GlobalVariableOp op) {
        return SPIRVLayoutInfoUtils::isLegalType(op.type());
      });

```


My base assumptions for this pass are:
1. SPIR-V dialect is legal and therefore all SPIR-V ops are legal.
2. FuncOp is legal, because SPIR-V dialect reuses it from standard ops.
3. spirv:GlobalVariableOp is dynamically legal, it's illegal if:
a. Has special type. b. Has special strorage class c. Does not have layout info.
After pattern is applied for illegal op it becomes legal.

The struct with custom layout is legal, because it has layout.
Here is a test:  
```
// CHECK: spv.globalVariable @customLayout : !spv.ptr<!spv.struct<f32 [64], i32 [128]>, Uniform>
  spv.globalVariable @customLayout : !spv.ptr<!spv.struct<f32 [64], i32 [128]>, Uniform>


```
by the way, I've just started to learn MLIR Legalization machinery, and it could be another and better way to apply a convertion for those ops, please fix me in this case.

Reply:
Ok, thanks for the clarification. Your understanding is how I understand as well. Since there is already a test, this seems OK to me. Please ignore my concern on this point!